File: app/src/main/java/com/pixieium/austtravels/auth/SignInActivity.java
package com.pixieium.austtravels.auth;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.pixieium.austtravels.R;
import com.pixieium.austtravels.databinding.ActivitySignInBinding;
import com.pixieium.austtravels.home.HomeActivity;

public class SignInActivity extends AppCompatActivity {

    private ActivitySignInBinding mBinding;

    FirebaseAuth mAuth;




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mBinding = ActivitySignInBinding.inflate(getLayoutInflater());
        View view = mBinding.getRoot();
        setContentView(view);

        mBinding.signin.setOnClickListener(view1 -> {
            Intent intent = new Intent(getBaseContext(), HomeActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
        });

        mAuth = FirebaseAuth.getInstance();

        mBinding.signup.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(getApplicationContext(),SignUpActivity.class);
                startActivity(intent);
            }
        });

        mBinding.signin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                signinUserbyEmail();
            }
        });


    }

    @Override
    protected void onStart() {
        super.onStart();
        FirebaseUser user = mAuth.getCurrentUser();
        if(user!=null)
        {
            Intent intent = new Intent(getApplicationContext(),HomeActivity.class);
            startActivity(intent);
        }
        else {

        }
    }




    private void signinUserbyEmail(){
        String email = mBinding.email.getEditText().getText().toString();
        String password = mBinding.password.getEditText().getText().toString();

        if(TextUtils.isEmpty(email)||TextUtils.isEmpty(password)) {
            if (TextUtils.isEmpty(email)) {
                mBinding.email.setError("Email cannot be empty");
            }
            if (TextUtils.isEmpty(password)) {
                mBinding.password.setError("Password cannot be empty");
            }
        }
        else{
            mAuth.signInWithEmailAndPassword(email, password)
                    .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull Task<AuthResult> task) {
                            if (task.isSuccessful()) {
                                // Sign in success, update UI with the signed-in user's information

                                FirebaseUser user = mAuth.getCurrentUser();
                                Intent intent = new Intent(getApplicationContext(),HomeActivity.class);
                                startActivity(intent);

                            } else {

                                Toast.makeText(getApplicationContext(), "Authentication failed.",
                                        Toast.LENGTH_SHORT).show();

                            }
                        }
                    });
        }



    }

}

File: app/src/main/java/com/pixieium/austtravels/auth/SignInActivity.kt
package com.pixieium.austtravels.auth

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase
import com.pixieium.austtravels.databinding.ActivitySignInBinding
import com.pixieium.austtravels.home.HomeActivity
import java.lang.Exception

class SignInActivity : AppCompatActivity() {

    private lateinit var mBinding: ActivitySignInBinding
    private lateinit var mAuth: FirebaseAuth

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = ActivitySignInBinding.inflate(layoutInflater)
        val view: View = mBinding.root
        setContentView(view)

        mAuth = FirebaseAuth.getInstance()
        mBinding.signup.setOnClickListener {
            val intent = Intent(applicationContext, SignUpActivity::class.java)
            startActivity(intent)
        }
        mBinding.signin.setOnClickListener {
            try {
                // hide the keyboard
                hideKeyboard()
            } catch (e: Exception) {
                e.printStackTrace()
                Toast.makeText(this, e.localizedMessage, Toast.LENGTH_SHORT).show()
            } finally {
                signInUserByEmail()
            }
        }
    }

    private fun hideKeyboard() {
        // hide the keyboard
        val imm: InputMethodManager =
            getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(mBinding.root.windowToken, 0)
    }

    private fun signInUserByEmail() {
        val email = mBinding.email.editText!!.text.toString()
        val password = mBinding.password.editText!!.text.toString()
        if (TextUtils.isEmpty(email) || TextUtils.isEmpty(password)) {
            if (TextUtils.isEmpty(email)) {
                mBinding.email.error = "Email cannot be empty"
            }
            if (TextUtils.isEmpty(password)) {
                mBinding.password.error = "Password cannot be empty"
            }
        } else {
            mAuth.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(
                    this
                ) { task ->
                    if (task.isSuccessful) {
                        // Sign in success, update UI with the signed-in user's information
                        checkVerifiedEmail()
                    }
                }
                .addOnFailureListener {
                    Toast.makeText(this, it.localizedMessage, Toast.LENGTH_SHORT).show()
                }
        }
    }

    private fun checkVerifiedEmail() {
        val currentUser: FirebaseUser? = FirebaseAuth.getInstance().currentUser

        if (currentUser?.isEmailVerified == true) {
            val intent = Intent(baseContext, HomeActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
            startActivity(intent)
        } else {
            val snakbar: Snackbar = Snackbar.make(
                mBinding.signinLayout,
                "Please verify your email",
                Snackbar.LENGTH_LONG
            )
                .setAction("Resend email", View.OnClickListener {
                    currentUser?.sendEmailVerification()?.addOnCompleteListener(this) { task ->
                        if (task.isSuccessful) {

                            Toast.makeText(
                                this@SignInActivity,
                                "Email verification link sent to your email",
                                Toast.LENGTH_SHORT
                            ).show()

                        }

                    }
                })

            snakbar.show()

            Firebase.auth.signOut()
        }
    }
}